cryptarchia/cryptarchia.py

```diff
@@ -118,37 +122,103 @@ def on_block(self, block: BlockHeader):
             self.forks.append(new_chain)
         # otherwise, we're missing the parent block
         # in that case, just ignore the block
-        
 
     # Evaluate the fork choice rule and return the block header of the block that should be the head of the chain
     def fork_choice(local_chain: Chain, forks: List[Chain]) -> BlockHeader:
         pass
-        
 
     def tip(self) -> BlockHeader:
         return self.fork_choice()
 
 
 @dataclass
 class Coin:
+    pk: int
     value: int
 
+
+@dataclass
+class LedgerState:
+    """
+    A snapshot of the ledger state up to some height
+    """
+
+    height: int = None
+    nonce: bytes = None
+    total_stake: int = None
+
+
+@dataclass
+class EpochState:
+    # for details of snapshot schedule please see:
+    # https://github.com/IntersectMBO/ouroboros-consensus/blob/fe245ac1d8dbfb563ede2fdb6585055e12ce9738/docs/website/contents/for-developers/Glossary.md#epoch-structure
+
+    # The stake distribution snapshot is taken at the beginning of the previous epoch
+    stake_distribution_snapshot: LedgerState
+
+    # The nonce snapshot is taken 7k/f slots into the previous epoch
+    nonce_snapshot: LedgerState
+
+    def stake_distribution(self) -> int:
+        """Returns the total stake that will be used to reletivize leadership proofs during this epoch"""
+        return self.stake_distribution_snapshot.total_stake
+
+    def nonce(self) -> bytes:
+        return self.nonce_snapshot.nonce
```

Comment:
are we going to need something more complicated than the total stake or were total stake requires some calculations? If not, can we simplify this and remove `LedgerState`? We probably can't take a stake distribution snapshot anyway if coins are private

Reply:
Whatever data is derived from the snapshots is not too important right now, the important thing is that they are old enough to be used in the current epoch safely and that the snapshots are taken in the correct order.

I.e. stake distribution snapshot is taken before nonce snapshot to prevent grinding.

> are we going to need something more complicated than the total stake or were total stake requires some calculations?

There will be more to this as we expand the leader proofs spec. For example we need to prove that a coin was present when the stake snapshot was taken.

>  If not, can we simplify this and remove LedgerState?

As I was reading the ouroborous glossary, it seemed like the idea of ledger state is quite fundamental, for example the genesis is not a block but a genesis ledger state: https://github.com/IntersectMBO/ouroboros-consensus/blob/fe245ac1d8dbfb563ede2fdb6585055e12ce9738/docs/website/contents/for-developers/Glossary.md#ledger-state

Reply:
We're definitely going to need this at some point, just wondering whether it's best to introduce it when we actually need it.

> the important thing is that they are old enough to be used in the current epoch safely and that the snapshots are taken in the correct order.

true

Reply:
> the idea of ledger state is quite fundamental

I think you convinced me it may be useful to have this concept from the start even though at the moment it's not particularly populated